encoders/firebase-decoders-json/src/json/java/com/google/firebase/decoders/json/JsonDataDecoderBuilderContext.java

```diff
@@ -50,26 +53,110 @@
   private <T> T decode(TypeToken<T> typeToken) throws IOException {
     if (typeToken instanceof TypeToken.ClassToken) {
       TypeToken.ClassToken<T> classToken = (TypeToken.ClassToken<T>) typeToken;
-      CreationContextImpl creationContext = decodeObject(classToken);
-      @SuppressWarnings("unchecked")
-      // Safe, because typeToken and TypeCreator always have same type parameter
-      TypeCreator<T> creator = (TypeCreator<T>) typeCreators.get(classToken);
-      if (creator == null)
-        throw new IllegalArgumentException(
-            "TypeCreator of " + classToken.getRawType() + " is not register.");
-      return (T) creator.create(creationContext);
-
+      return decodeClassToken(classToken);
     } else if (typeToken instanceof TypeToken.ArrayToken) {
-      // TODO: Change typeParameter T in ArrayToken<T> to represent component type.
-      /**
-       * reader.beginArray(); List<T> l = new LinkedList<>(); while reader.hasNext: T val =
-       * decode(arrayToken.getComponentType()); l.add(val); reader.endArray(); return l.toArray();
-       */
+      TypeToken.ArrayToken<T> arrayToken = (TypeToken.ArrayToken<T>) typeToken;
+      return (T) decodeArrayToken(arrayToken);
     }
     return null;
   }
 
-  private <T> CreationContextImpl decodeObject(TypeToken.ClassToken<T> classToken)
+  private <T> T decodeClassToken(TypeToken.ClassToken<T> classToken) throws IOException {
+    if (classToken.getRawType().isPrimitive()) {
+      return decodePrimitive(classToken);
+    } else if (isSingleValue(classToken)) {
+      return decodeSingleValue(classToken);
+    } else {
+      return decodeObject(classToken);
+    }
+  }
+
+  private <T, E> T decodeArrayToken(TypeToken.ArrayToken<T> arrayToken) throws IOException {
```

Comment:
What's the purpose of the `E` type parameter? can you use `Object` in `TypeToken` below instead?

Reply:
My initial thought was to align with `convertGenericListToArray` since `convertGenericListToArray` need `E` to create `E[]` generic array to assign value to its elements.

Reply:
But it's a generic method, maybe I don't need to align?